src/cockpit/peer.py

```diff
@@ -142,3 +142,41 @@ def do_channel_control(self, channel: str, command: str, message: Dict[str, obje
 
     def do_channel_data(self, channel: str, data: bytes) -> None:
         self.write_channel_data(channel, data)
+
+
+class PeerRoutingRule(RoutingRule, PeerStateListener):
+    predicate: Callable[[Dict[str, object]], bool]
+    args: List[str]
+    env: Dict[str, str]
+    peer: Optional[Peer]
+
+    def __init__(self, router: Router, args: List[str], env: Dict[str, str], predicate: Callable[[Dict[str, object]], bool]):
+        super().__init__(router)
+
+        self.args = args
+        self.env = env
+        self.predicate = predicate
+        self.peer = None
+
+    def apply_rule(self, options: Dict[str, object]) -> Optional[Peer]:
+        if not self.predicate(options):
+            return None
```

Comment:
This needs some documenation what apply_rule() actually does and what it returns. Currently it *both* has a side effect (changing its internal state `self.peer`) *and* a return value (returning a "peer"). It's also not actually being called anywhere, so I can't infer how it's supposed to be used.

Reply:
Ah, this overrides `RoutingRule.apply_rule`, that has a docstring.